### ISSUE TITLE
Interpolate variables in terraform backend config

### ISSUE THREAD

**Issue by roderickrandolph** [hashicorp/terraform#12270]
### Terraform Version
Terraform v0.9.0-beta1

### Affected Resource(s)
core

### Terraform Configuration Files
```hcl
provider "aws" { region = "us-east-1" }

terraform {
  backend "s3" {
    bucket = "my-bucket-${var.aws_region}"
    key = "${var.stack_name}.tfstate"
    region = "${var.aws_region}"
    encrypt = "1"
  }
}
variable "aws_region" { default = "us-east-1" }
variable "stack_name" { default = "my-stack" }
```

### Debug Output
N/A

### Panic Output
N/A

### Expected Behavior
Terraform will interpolate all variables provided in the backend configuration (i.e. `${var.aws_region}` and `${var.stack_name}`)

### Actual Behavior
Terraform treats `${var.stack_name}` and `${var.aws_region}` as literal strings causing a terraform plan/apply to fail. 
```
Error loading state: InvalidEndpointURL: invalid endpoint uri
caused by: parse https://s3.${var.aws_region}.amazonaws.com/{Bucket}/{Key+}: invalid character "{" in host name
```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform plan`

**Comment by mitchellh** [comment]
This isn't allowed and beta 2 will properly give you an error message. Rather than re-explain, I'll link you to the PR with the discussion we had! https://github.com/hashicorp/terraform/pull/12067 We do want you to be able to parameterize backends but for now that'll be restricted to partial configs with `-backend-config` on `init`.

**Comment by roderickrandolph** [comment]
Thank you. I didn't realize there was a previous discussion on this topic. I actually looked before opening this issue too.

General feedback:
From a UX perspective the previous behavior of passing in multiple `-backend-config` as K/V pairs on the command line was preferred over generating a separate throw away config file for each environment stack. Overall, I like the new `terraform init` direction but the partial backend config file seems like a step backwards from a UX perspective.

Also, given the argument name I initially assumed `-backend-config` for `terraform init` would behave like it did for `terraform remote config` and accept K/V pairs until after reading the docs more closely I realized it takes a file path instead. Others will probably have similar experiences too as they make the transition to 0.9. Perhaps a different argument name could reduce the confusion. Just a thought.
